flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingMetricEvaluator.java

```diff
@@ -154,6 +153,56 @@ private Map<ScalingMetric, EvaluatedScalingMetric> evaluateMetrics(
         return evaluatedMetrics;
     }
 
+    private static EvaluatedScalingMetric evaluateTpr(
+            SortedMap<Instant, CollectedMetrics> metricsHistory,
+            JobVertexID vertex,
+            Map<ScalingMetric, Double> latestVertexMetrics,
+            Configuration conf) {
+
+        var busyTimeTprAvg = getAverage(TRUE_PROCESSING_RATE, vertex, metricsHistory);
+        var observedTprAvg =
+                getAverage(
+                        OBSERVED_TPR,
+                        vertex,
+                        metricsHistory,
+                        conf.get(AutoScalerOptions.OBSERVED_TPR_MIN_OBSERVATIONS));
+
+        var tprMetric = selectTprMetric(vertex, conf, busyTimeTprAvg, observedTprAvg);
+        return new EvaluatedScalingMetric(
+                latestVertexMetrics.getOrDefault(tprMetric, Double.NaN),
+                tprMetric == OBSERVED_TPR ? observedTprAvg : busyTimeTprAvg);
+    }
+
+    private static ScalingMetric selectTprMetric(
+            JobVertexID jobVertexID,
+            Configuration conf,
+            double busyTimeTprAvg,
+            double observedTprAvg) {
+
+        if (Double.isInfinite(busyTimeTprAvg) || Double.isNaN(busyTimeTprAvg)) {
+            return OBSERVED_TPR;
+        }
+
+        if (Double.isNaN(observedTprAvg)) {
+            return TRUE_PROCESSING_RATE;
+        }
+
+        double switchThreshold = conf.get(AutoScalerOptions.OBSERVED_TPR_SWITCH_THRESHOLD);
+        // If we could measure the observed tpr we decide whether to switch to using it
+        // instead of busy time based on the error / difference between the two
+        if (busyTimeTprAvg > observedTprAvg * (1 + switchThreshold)) {
```

Comment:
Should we add a config-based switch here to turn this on / off?

Reply:
I am not completely sure, the user could set a large lag threshold / switch threshold to effectively disable the feature. But in general if this works correctly I am not sure why anyone would want to disable it, instead of simply tuning the threshold maybe.

Reply:
old PR, but fwiw I am now seeing a large _under_estimation of capacity:

`Using observed tpr 916327.7340000003 for 1385b0b55fdb909b073b014d9c682612 as busy time based seems too large (2018541.9166666663)`

In fact the actual capacity is indeed around 2M, if not higher. We are on 1.7, but perhaps 1.8 will have improved its observed TPR due to [this enhancement](https://flink.apache.org/2024/03/21/apache-flink-kubernetes-operator-1.8.0-release-announcement/#improved-accuracy-of-autoscaling-metrics)?

Reply:
The observed true processing rate is simply measured based on the num records processed when the job is catching up. Why do you think that your job as 2M/sec processing capacity?

Maybe when the job is catching up it cannot go 2M / sec ?

Reply:
It's possible if I am misunderstanding something!

During the same time that log was printed during a downscale, I have metrics showing that actual `flink_taskmanager_job_task_operator_numRecordsIn` (per sec) was above 1.2M / sec for 15 minutes before and after, and peaked at about 2.8M for around 3 minutes. The peak was due to the rescale.

It caused it to severely underestimate the capacity, scale way back up (higher than it was before the downscale), and then eventually settle back down to where it tried to go initially. At that point it was no longer seeing the discrepancy between observed and busy-time-based.

I am guessing (but probably wrong 😄) that it has more to do with _when_ it's measuring the "observed" TPR than it being wrong? Like it perhaps misses the window where it's happening. Our restart time is 30s, stabilization time 5m, window 15m. The whole recovery lasted maybe 3m. If those times are not aligned with reality (it likely restarted much quicker than 30s) then could it cause it to miss the peak observation TPR?

Reply:
Observed TPR is a bit of outlier metric for a "good" reason. It is actually computed regardless of the stabilisation time as you can see [here](https://github.com/apache/flink-kubernetes-operator/blob/main/flink-autoscaler/src/main/java/org/apache/flink/autoscaler/metrics/ScalingMetrics.java#L104). There is a config threshold for the observed tpr computation which sets a minimum source lag when this can be computed (set at 30 seconds by default)

The rationale is that you can only really observe the true processing rate when the job is catching up (going full speed). So that's the reason here, maybe you can set that threshold a bit higher (1m) which would mean that we measure only when there is at least 1 minute lag. Higher threshold will mean more accurate measurement when it's triggered but less overall datapoints.

Reply:
If your job catches up super quick then you may find anomalies for these measurements because then you will likely only have 1-2 datapoints at most and that can be off arbitrarily

Reply:
Great idea, thanks for the tip! I saw that config setting but did not understand what it was actually used for. I will experiment with increasing that setting - our jobs are all over the place in how quickly they recover (some are a few seconds, some are several minutes) but I would bias towards accuracy in order to avoid the flapping in our case.